### PR TITLE
Add support for dpctl.dparray.

### PR DESCRIPTION
**Pull request by PokhodenkoSA**
Moved code from https://github.com/IntelPython/numba/pull/116.
- [x] Update with main
- [x] Transform unit tests `test_usmarray.py`
- [x] Need to get rid of relative paths in `setup.py`

### CODE REVIEW

**Review by DrTodd13 — APPROVED** [review]
Looks correct at first glance.  I didn't look to see why the tests were failing but that would be expected even if this is using the dpctl dparray branch since we removed the section of code there that looks for Numba meminfo objects.  Once this is merged then I will add the code here to monkey-patch the meminfo objects with sycl_usm_array_interface.

File: numba_dppy/numpy_usm_shared.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    return tuple([z if z != types.Array else UsmSharedArrayType for z in x])
+
+
+def numba_register_lower_builtin():
+    todo = []
+    todo_builtin = []
+    todo_getattr = []
+
+    # For all Numpy identifiers that have been registered for typing in Numba...
+    # this registry contains functions, getattrs, setattrs, casts and constants...need to do them all? FIX FIX FIX
```
Comment (diptorupd) [inline numba_dppy/numpy_usm_shared.py:191]: Please convert this comment into a ticket.

**Comment by PokhodenkoSA** [comment]
> This code was moved to `dpctl.dptensor`.

Is it relate to some line?

**Comment by PokhodenkoSA** [comment]
This PR uses https://github.com/IntelPython/dpctl/pull/214.

**Comment by PokhodenkoSA** [comment]
Assertion error in last case:
```
print("------------------- Testing Numba usmarray.T")
sys.stdout.flush()
dp4 = f12(dp3)
print("dp4:", dp4, type(dp4))
assert isinstance(dp4, usmarray.ndarray)
assert usmarray.has_array_interface(dp4)
del dp3
del dp4
```

```
<class 'numpy.ndarray'>
Traceback (most recent call last):
  File "numba_dppy/tests/test_usmarray.py", line 220, in <module>
    assert isinstance(dp4, usmarray.ndarray)
AssertionError
```

File: setup.py [hunk 2]
```diff
@@ -3,10 +3,29 @@
...
+import sys
+
+
+def find_numba():
+    numba_dir = ""
+    sys_packages = sys.path
+    for pcg in sys_packages:
+        if pcg.find("/numba-0") != -1:
+            numba_dir = pcg
+    return numba_dir
```
Comment (PokhodenkoSA) [inline setup.py:15]: @1e-to please use `numba.core.extending.include_path()`

Comment (PokhodenkoSA) [reply setup.py:15]: Docs about it - https://numba.pydata.org/numba-doc/dev/developer/numba-runtime.html?highlight=include_path#using-the-nrt-from-c-code

**Comment by PokhodenkoSA** [comment]
3 tests failed and marked as expected failure:
- test_numba_usmarray_mul_add - strange, it passed if runs not in butch
- test_numba_usmarray_constructor_from_numpy_ndarray - was commented previously
- test_numba_usmarray_T - did not work as I started

**Comment by DrTodd13** [comment]
> * `test_numba_usmarray_constructor_from_numpy_ndarray` - was commented previously

I think we agreed that such implicit conversions would not be allowed so this test can be removed.

**Comment by PokhodenkoSA** [comment]
I think I have found why CI fails. When installed from conda package in clean environment extension fails to activate. It writes that dpCtl requires cython.
Do not know why dpCtl requires Cython but when it installed extension activates and passes.

**Comment by PokhodenkoSA** [comment]
Current problem on Windows:
```
test_dppy_fallback_false (numba_dppy.tests.test_controllable_fallback.TestDPPYFallback) ...
...
Did not find DPCTLQueueMgr_GetCurrentQueue in libDPCTLSyclInterface.dll
```

File: numba_dppy/dppy_rt.c [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        char *malloc_name = "DPCTLmalloc_shared";
+        char *free_name = "DPCTLfree_with_queue";
+        char *get_queue_name = "DPCTLQueueMgr_GetCurrentQueue";
+        char *free_queue_name = "DPCTLQueue_Delete";
+
+        void *sycldl = dlopen(lib_name, RTLD_NOW);
+        assert(sycldl != NULL);
+        internal_allocator = (NRT_external_malloc_func)dlsym(sycldl, malloc_name);
+        usmarray_allocator.malloc = save_queue_allocator;
+        if (usmarray_allocator.malloc == NULL) {
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_rt.c:55]: This check always passed. Should it be `if (internal_allocator == NULL) {` ?

File: numba_dppy/dppy_rt.c [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            printf("Did not find %s in %s\n", malloc_name, lib_name);
+            exit(-1);
+        }
+
+        usmarray_allocator.realloc = NULL;
+
+        internal_free = (NRT_external_free_func)dlsym(sycldl, free_name);
+        usmarray_allocator.free = save_queue_deallocator;
+        if (usmarray_allocator.free == NULL) {
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_rt.c:64]: This check is also strange for me.

File: numba_dppy/dppy_rt.c [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        char *malloc_name = "DPCTLmalloc_shared";
+        char *free_name = "DPCTLfree_with_queue";
+        char *get_queue_name = "DPCTLQueueMgr_GetCurrentQueue";
+        char *free_queue_name = "DPCTLQueue_Delete";
+
+        HMODULE sycldl = LoadLibrary(lib_name);
+        assert(sycldl != NULL);
+        internal_allocator = (NRT_external_malloc_func)GetProcAddress(sycldl, malloc_name);
+        usmarray_allocator.malloc = save_queue_allocator;
+        if (usmarray_allocator.malloc == NULL) {
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_rt.c:92]: The same strange check.

File: numba_dppy/dppy_rt.c [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            printf("Did not find %s in %s\n", malloc_name, lib_name);
+            exit(-1);
+        }
+
+        usmarray_allocator.realloc = NULL;
+
+        internal_free = (NRT_external_free_func)GetProcAddress(sycldl, free_name);
+        usmarray_allocator.free = save_queue_deallocator;
+        if (usmarray_allocator.free == NULL) {
```
Comment (PokhodenkoSA) [inline numba_dppy/dppy_rt.c:101]: The same strange check.

File: numba_dppy/dppy_rt.c (hunk 3 above)
Comment (DrTodd13) [reply numba_dppy/dppy_rt.c:55]: Yes, it should be internal_allocator.

File: numba_dppy/dppy_rt.c (hunk 4 above)
Comment (DrTodd13) [reply numba_dppy/dppy_rt.c:64]: Yes, this should also check internal_free.

File: numba_dppy/dppy_rt.c (hunk 5 above)
Comment (DrTodd13) [reply numba_dppy/dppy_rt.c:92]: Same as above.

File: numba_dppy/dppy_rt.c (hunk 6 above)
Comment (DrTodd13) [reply numba_dppy/dppy_rt.c:101]: Same fix as above.

**Review by DrTodd13 — CHANGES_REQUESTED** [review]

File: numba_dppy/dppy_rt.c (hunk 4 above)
Comment (PokhodenkoSA) [reply numba_dppy/dppy_rt.c]: Fixed it all.

**Comment by PokhodenkoSA** [comment]
I would like to merge it. @DrTodd13 what is not ready yet?

**Comment by DrTodd13** [comment]
> I would like to merge it. @DrTodd13 what is not ready yet?
> 
> I would like to merge it before #135 - black formatter. However I can do it in any order.

There are some parts of the USM array attribute class that use bound_function that don't seem to work with the current configuration and I'm not sure why.  I say we merge what we currently have while I investigate this issue I described and we can have another PR for that.

File: numba_dppy/numpy_usm_shared.py [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+# # Load the symbols from the DPPL Sycl library.
+# llb.load_library_permanently(sycl_mem_lib)
+
+import dpctl
+from dpctl.memory import MemoryUSMShared
+import numba_dppy._dppy_rt
+
+# functions_list = [o[0] for o in getmembers(np) if isfunction(o[1]) or isbuiltin(o[1])]
+# class_list = [o for o in getmembers(np) if isclass(o[1])]
+
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:50]:
```suggestion
```

File: numba_dppy/numpy_usm_shared.py [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+# Register the helper function in dppl_rt so that we can insert calls to them via llvmlite.
+for py_name, c_address in numba_dppy._dppy_rt.c_helpers.items():
+    llb.add_symbol(py_name, c_address)
+
+
+# This class creates a type in Numba.
+class UsmSharedArrayType(types.Array):
+    def __init__(
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:58]:
```suggestion
class UsmSharedArrayType(types.Array):
    """This class creates a type in Numba."""
    def __init__(
```

File: numba_dppy/numpy_usm_shared.py [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+    )
+    fn.return_value.add_attribute("noalias")
+    if isinstance(align, builtins.int):
+        align = context.get_constant(types.uint32, align)
+    else:
+        assert align.type == u32, "align must be a uint32"
+    return builder.call(fn, [size, align, ext_allocator])
+
+
+registered = False
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:155]:
```suggestion
_registered = False
```

File: numba_dppy/numpy_usm_shared.py [hunk 10]
```diff
@@ -0,0 +1,713 @@
...
+        if isinstance(val, (ftype, bftype)):
+            # If we have overloaded that function in the usmarray module (always True right now)...
+            if val.__name__ in functions_list:
+                todo.append(ig)
+        if isinstance(val, type):
+            if isinstance(typ, numba.core.types.functions.Function):
+                todo.append(ig)
+            elif isinstance(typ, numba.core.types.functions.NumberClass):
+                pass
+                #todo_classes.append(ig)
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:274]:
```suggestion
```

File: numba_dppy/numpy_usm_shared.py [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+
+        def generic_impl(self):
+            original_typer = self.__class__.original.generic(self.__class__.original)
+            ot_argspec = inspect.getfullargspec(original_typer)
+            astr = argspec_to_string(ot_argspec)
+
+            typer_func = """def typer({}):
+                                original_res = original_typer({})
+                                if isinstance(original_res, types.Array):
+                                    return UsmSharedArrayType(dtype=original_res.dtype, ndim=original_res.ndim, layout=original_res.layout)
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:332]: Long line.

File: numba_dppy/numpy_usm_shared.py [hunk 12]
```diff
@@ -0,0 +1,713 @@
...
+
+        new_usmarray_template = type(
+            class_name,
+            (tgetattr,),
+            {"set_class_vars": set_key, "__getattribute__": getattr_impl},
+        )
+
+        new_usmarray_template.set_class_vars(UsmSharedArrayType)
+        templates_registry.register_attr(new_usmarray_template)
+    """
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:416]: Commented code.

File: numba_dppy/numpy_usm_shared.py [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+        else:
+            raise TypeError("take(%s) not supported for %s" % argty)
+        return sig
+
+    def generic_resolve(self, ary, attr):
+        # Resolution of other attributes, for record arrays
+        if isinstance(ary.dtype, types.Record):
+            if attr in ary.dtype.fields:
+                return ary.copy(dtype=ary.dtype.typeof(attr), layout='A')
+"""
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:685]: Commented code.

File: numba_dppy/tests/test_usmarray.py [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+def numba_mul_usmarray_asarray(a, b):  # a is usmarray, b is numpy
+    return a * usmarray.asarray(b)
+
+
+# @numba.njit()
+# def f7(a):  # a is usmarray
+#     # implicit conversion of a to numpy.ndarray
+#     b = numpy.ones(10)
+#     c = a * b
+#     d = a.argsort()  # with no implicit conversion this fails
```
Comment (PokhodenkoSA) [inline numba_dppy/tests/test_usmarray.py:33]: Will remove it.

**Comment by PokhodenkoSA** [comment]
@DrTodd13 I have find interesting tests dependency and possibly an issue.
```python
assert isinstance(numba_mul_add(numpy.ones(10)), numpy.ndarray)
assert not isinstance(numba_mul_add(usmarray.ones(10)), usmarray.ndarray)
```
Function was compiled and cashed as something which takes `numpy.ndarray` because it was passed first time.

File: numba_dppy/tests/test_usmarray.py (hunk 14 above)
Comment (PokhodenkoSA) [inline numba_dppy/tests/test_usmarray.py:33]:
```suggestion
```

File: numba_dppy/numpy_usm_shared.py [hunk 15]
```diff
@@ -0,0 +1,727 @@
...
+# dprint("sycl_mem_lib:", sycl_mem_lib)
+# # Load the symbols from the DPPL Sycl library.
+# llb.load_library_permanently(sycl_mem_lib)
+
+import dpctl
+from dpctl.memory import MemoryUSMShared
+import numba_dppy._dppy_rt
+
+# functions_list = [o[0] for o in getmembers(np) if isfunction(o[1]) or isbuiltin(o[1])]
+# class_list = [o for o in getmembers(np) if isclass(o[1])]
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:55]:
```suggestion
```

**Comment by PokhodenkoSA** [comment]
@DrTodd13 
I have made formatting for the code with `black`. If you will face merge conflicts do following:
```bash
pip install black
black .
```

File: numba_dppy/numpy_usm_shared.py (hunk 8 above)
Comment (diptorupd) [reply numba_dppy/numpy_usm_shared.py:64]: If we are making this change, let's make the docstring more meaning full.

File: numba_dppy/numpy_usm_shared.py (hunk 15 above)
Comment (DrTodd13) [reply numba_dppy/numpy_usm_shared.py:56]: Done.

File: numba_dppy/tests/test_usmarray.py (hunk 14 above)
Comment (DrTodd13) [reply numba_dppy/tests/test_usmarray.py:33]: Done.

File: numba_dppy/numpy_usm_shared.py (hunk 13 above)
Comment (DrTodd13) [reply numba_dppy/numpy_usm_shared.py]: Code is active again.

File: numba_dppy/numpy_usm_shared.py (hunk 12 above)
Comment (DrTodd13) [reply numba_dppy/numpy_usm_shared.py:453]: Done.

File: numba_dppy/numpy_usm_shared.py (hunk 8 above)
Comment (DrTodd13) [reply numba_dppy/numpy_usm_shared.py:65]: Done.

File: numba_dppy/numpy_usm_shared.py (hunk 9 above)
Comment (DrTodd13) [reply numba_dppy/numpy_usm_shared.py:171]: Done.

File: numba_dppy/numpy_usm_shared.py (hunk 10 above)
Comment (DrTodd13) [reply numba_dppy/numpy_usm_shared.py]: Done.

File: numba_dppy/numpy_usm_shared.py (hunk 11 above)
Comment (DrTodd13) [reply numba_dppy/numpy_usm_shared.py:369]: Done.

**Comment by PokhodenkoSA** [comment]
@DrTodd13 is it ready to be merged?

**Comment by DrTodd13** [comment]
> @DrTodd13 is it ready to be merged? How can I help with it?

Yes.  We can merge.  There are some known issues to be fixed but let's get the stuff that is working checked in and fix the rest later.

File: numba_dppy/numpy_usm_shared.py [hunk 16]
```diff
@@ -0,0 +1,728 @@
...
+
+
+def numba_register_lower_builtin():
+    todo = []
+    todo_builtin = []
+    todo_getattr = []
+    todo_array_member_func = []
+
+    # For all Numpy identifiers that have been registered for typing in Numba...
+    # this registry contains functions, getattrs, setattrs, casts and constants...need to do them all? FIX FIX FIX
```
Comment (PokhodenkoSA) [inline numba_dppy/numpy_usm_shared.py:232]:
```suggestion
    # For all Numpy identifiers that have been registered for typing in Numba...
    # this registry contains functions, getattrs, setattrs, casts and constants...
```
https://github.com/IntelPython/numba-dppy/issues/142

**Comment by PokhodenkoSA** [comment]
@diptorupd @DrTodd13 
Could somebody of you please approve this PR ?

**Pull request merged**